Upgrade to 1.92.1

kcbanner opened the pull request:
- Upgrade to 1.92.1
- Add TextureRef and update related APIs
- Set IMGUI_DISABLE_OBSOLETE_FUNCTIONS and remove usage of deprecated functions
- Fixup deprecated API usages (callconv, alignment)

Remaining TODOs:
- [x] Update test engine
- [x] Update node editor

kcbanner commented:
This is ready for review, but currently the build.zig.zon points to my forks for these changes: https://github.com/zig-gamedev/zsdl/pull/24 and https://github.com/zig-gamedev/zgpu/pull/18.

I'll update the build.zig.zon once those are merged.

a-day-old-bagel commented:
I merged some of this into my own zgui fork so that I could test it against the zgpu changes I've been making.

There are a few things that might have to be addressed later, but there was one thing I noticed which surely can be fixed right away - in zgui.cpp, there are some lingering usages of ImTextureID which I think should have been changed to ImTextureRef, but were maybe overlooked.

I'll just link to the commit where I fixed them on my own fork to show you what I'm talking about:

https://github.com/a-day-old-bagel/zgui/commit/896022e0f7c2bd3faa45446d8972a9ff7ac99b03

If these changes aren't made, using the affected functions will crash due to ABI mismatches.

a-day-old-bagel commented:
I have another small issue to bring up - not fatal like an ABI mismatch or anything, but some of the new imgui backend code is triggering the memory leak / static memory warning in gui.zig when an application exits. It's a case of "guess when this static memory gets a destructor call" that everyone loves from C++.

[The static variable to blame is here in imgui_impl_glfw.cpp](https://github.com/kcbanner/zgui/blob/9c817dc2d88340ef1009b2e23539f0186313292a/libs/imgui/backends/imgui_impl_glfw.cpp#L181)

[The memory check, for reference, is here](https://github.com/kcbanner/zgui/blob/9c817dc2d88340ef1009b2e23539f0186313292a/src/gui.zig#L78)

[And this is the change I've tentatively made to deal with it for my own code](https://github.com/a-day-old-bagel/zgui/commit/3bd1cb00dc4fdbe2be3e33224abd340e7dd71857), but there may be a nicer way to do it that doesn't require patching the imgui code like that, not sure.

This could probably be filed as a "don't care, won't fix," but I hate seeing warnings on exit even if they don't matter functionally, so I just wanted to note it here.

kcbanner commented:
> There are a few things that might have to be addressed later, but there was one thing I noticed which surely can be fixed right away - in zgui.cpp, there are some lingering usages of ImTextureID which I think should have been changed to ImTextureRef, but were maybe overlooked.

Fixed these up, good catch.

> I have another small issue to bring up - not fatal like an ABI mismatch or anything, but some of the new imgui backend code is triggering the memory leak / static memory warning in gui.zig when an application exits. It's a case of "guess when this static memory gets a destructor call" that everyone loves from C++.

I resolved this with a call to `clear()` in `ImGui_ImplGlfw_Shutdown`, instead of a manual destructor call. 

I also switched my project from using `.no_backend` to `.gflw` (I was manually including the h/cpp before), and noticed that when using the shared library option the extern functions weren't found during linking. This was for two reasons - their names were being mangled and they weren't being imported from the dll. I resolved this by adding `extern "C"` and using the `@extern` syntax instead of `extern fn`. I had made a similar change to this in `zmesh` previously to solve a similar linking issue. I think this is a good solution and the same change should be made to the other backends. 

I'll continue with that solution (update other backends to use `@extern`) if that looks good to you guys.

a-day-old-bagel commented:
> I resolved this with a call to `clear()` in `ImGui_ImplGlfw_Shutdown`, instead of a manual destructor call.


Oh nice, that's a much better solution to the memory warning. Thanks for fixing that.

hazeycode commented:
Sorry but I had to revert this because it broke the opengl backend and the newer wgpu backend seems to incompatible with the version of dawn we're currently using in zgpu
